### PR TITLE
Populate a JsonHeaders.RESOLVABLE_TYPE on reply

### LINKED ISSUE #277: amqp cafe application sample broken due to ClassCastException
When I run the amqp cafe application sample I get a 

`ClassCastException: org.springframework.integration.samples.cafe.OrderItem cannot be cast to org.springframework.integration.samples.cafe.Drink`

in the `Waiter.prepareDelivery` method. I tried both the master and the 5.2.x branch. 

I looked through the code and the configuration xml files, but the reason is escaping me. Everything seems to be correctly wired up. 

`<int:json-to-object-transformer id="json-to-drink" input-channel="preparedJsonDrinks" output-channel="preparedDrinks" type="org.springframework.integration.samples.cafe.Drink"/>`

should deserialize incoming messages to `Drink` objects, yet they get deserialized to `OrderItem`.

Attached is a full console output & stack trace.
[trace.txt](https://github.com/spring-projects/spring-integration-samples/files/4113673/trace.txt)

### PR DESCRIPTION
**Pull request by artembilan**
Fixes https://github.com/spring-projects/spring-integration-samples/issues/277

In the `AbstractAmqpOutboundEndpoint` the `JsonHeaders.RESOLVABLE_TYPE`
from request message is copied to reply message making inconsistency downstream.
The `JsonToObjectTransformer` consults first a `JsonHeaders.RESOLVABLE_TYPE`
and deserialize payload to wrong type

* Fix `DefaultAmqpHeaderMapper` to populate a `JsonHeaders.RESOLVABLE_TYPE`
alongside with other `JsonHeaders` populated from the reply AMQP message.
This way a `JsonHeaders.RESOLVABLE_TYPE` from request message won't have effect
* To get access to classes, supply `AbstractHeaderMapper` with a bean factory `ClassLoader`
* Introduce a couple utility methods into `JsonHeaders` for building a `ResolvableType`

**Cherry-pick to 5.2.x**

### CODE REVIEW

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java [hunk 1]
```diff
@@ -109,40 +109,43 @@ protected DefaultAmqpHeaderMapper(String[] requestHeaderNames, String[] replyHea
...
+					.acceptIfNotNull(AmqpHeaders.RECEIVED_EXCHANGE, amqpMessageProperties.getReceivedExchange(),
+							headers::put)
+					.acceptIfHasText(AmqpHeaders.RECEIVED_ROUTING_KEY, amqpMessageProperties.getReceivedRoutingKey(),
+							headers::put)
+					.acceptIfNotNull(AmqpHeaders.REDELIVERED, amqpMessageProperties.isRedelivered(), headers::put)
+					.acceptIfNotNull(AmqpHeaders.REPLY_TO, amqpMessageProperties.getReplyTo(), headers::put)
+					.acceptIfNotNull(AmqpHeaders.TIMESTAMP, amqpMessageProperties.getTimestamp(), headers::put)
+					.acceptIfHasText(AmqpHeaders.TYPE, amqpMessageProperties.getType(), headers::put)
+					.acceptIfHasText(AmqpHeaders.RECEIVED_USER_ID, amqpMessageProperties
+							.getReceivedUserId(), headers::put);
```
Comment (garyrussell) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Formatting inconsistent with the two changes above. This one looks more correct to me.

**Review by garyrussell — CHANGES_REQUESTED** [review]
Just one nit-pick.

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java (hunk 1 above)
Comment (artembilan) [reply spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Really? And I don't like breaking by dot and leaving a variable on the line with other calls. 😄

Comment (garyrussell) [reply spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: I don't really mind either way; but I think it should be consistent.

Comment (artembilan) [reply spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Yes, please, see the latest commit.

Thanks

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Cherry-picked to 5.2.x as 75476bd365149199fd3d1842df85724f360bf6df

**Pull request merged**